compiler/rustc_attr_parsing/src/attributes/used.rs

```diff
@@ -0,0 +1,111 @@
+use rustc_attr_data_structures::lints::AttributeLintKind;
+use rustc_attr_data_structures::{AttributeKind, UsedBy};
+use rustc_feature::template;
+use rustc_macros::Diagnostic;
+use rustc_session::parse::feature_err;
+use rustc_span::{Span, sym};
+
+use crate::attributes::{AcceptMapping, AttributeParser};
+use crate::context::{FinalizeContext, Stage};
+use crate::parser::ArgParser;
+
+#[derive(Diagnostic)]
+#[diag(attr_parsing_used_compiler_linker)]
+pub(crate) struct UsedCompilerLinker {
+    #[primary_span]
+    pub spans: Vec<Span>,
+}
+
+#[derive(Default)]
+pub(crate) struct UsedParser {
+    first_compiler: Option<Span>,
+    first_linker: Option<Span>,
+    result: Option<(UsedBy, Span)>,
+}
+
+// A custom `AttributeParser` is used rather than a Simple attribute parser because
+// - Specifying two `#[used]` attributes is a warning (but will be an error in the future)
+// - But specifying two conflicting attributes: `#[used(compiler)]` and `#[used(linker)]` is already an error today
+// We can change this to a Simple parser once the warning becomes an error
```

Comment:
Is this comment correct, that this logic can't be expressed with the `Simple` attribute parser?

Reply:
hm that's a little sad that we can't nicely represent that.